### PR TITLE
use Hyper as a HTTP backend

### PR DESCRIPTION
**Pull request by bagder**
Status: https://github.com/curl/curl/wiki/Hyper

Made a draft PR early to make sure this work doesn't break the "normal" builds. **expect rebases!**

- [X] make configure find Hyper
- [X] initial code that creates and issues a request with Hyper
- [X] use the correct method, path and default headers
- [X] docs explaining how to build with Hyper
- [X] receive/store response correctly
- [X] debug callback for received content
- [X] pass received data and headers to correct callbacks
- [X] no memory-leaks
- [X] works with HTTPS (if done with HTTP/1)
- [X] tell Hyper to not lowercase headers
- [X] debug callback for sent headers
- [X] supports custom request headers
- [X] HTTP/2
- [X] curl -I (`HEAD`) works over HTTP(S) and h1/h2
- [X] over proxies
- [X] referer (`-e` )
- [X] redirects
- [X] cookies
- [X] test suite "Hyper compatible"
- [X] Basic auth headers
- [X] Basic proxy auth headers
- [X] simple POSTs work
- [X] handle POST/PUT
- [X] multi-part formpost
- [X] send custom request headers without values
- [X] pass on the "status reason" from the HTTP/1 status line
- [X] provide Hyper  version in `curl_version_info` function
- [X] handle `CURL_HTTP_VERSION_1_0`
- [X] chunked-encoded h1 "upload"
- [X] debug callback for sent content
- [ ] multiplexed HTTP/2
- [ ] h2 `Upgrade:`
- [ ] HTTP auth
- [ ] pausing transfers
- [ ] co-exist with a HTTP/3 build
- [ ] receiving HTTP/1 trailers
- [ ] sending HTTP/1 trailers
- [ ] accept-encoding
- [ ] transfer encoding

### CODE REVIEW

File: lib/setopt.c [hunk 1]
```diff
@@ -863,14 +863,18 @@ CURLcode Curl_vsetopt(struct Curl_easy *data, CURLoption option, va_list param)
...
     if(arg >= CURL_HTTP_VERSION_2)
       return CURLE_UNSUPPORTED_PROTOCOL;
 #else
     if(arg >= CURL_HTTP_VERSION_LAST)
       return CURLE_UNSUPPORTED_PROTOCOL;
     if(arg == CURL_HTTP_VERSION_NONE)
       arg = CURL_HTTP_VERSION_2TLS;
+#endif
+#ifdef USE_HYPER
+    if(arg == CURL_HTTP_VERSION_1_0)
```
Comment (seanmonstar) [inline lib/setopt.c:876]: I've just added `hyper_request_set_version(req, HYPER_HTTP_VERSION_1_0)`, which will enforce 1.0 semantics on a 1.x connection.

Comment (bagder) [reply lib/setopt.c:876]: Lovely, I'm adapting and will backpedal this commit. Thanks!

File: docs/libcurl/symbols-in-versions [hunk 2]
```diff
@@ -867,6 +867,7 @@ CURLVERSION_EIGHTH              7.72.0
 CURLVERSION_FIFTH               7.57.0
 CURLVERSION_FIRST               7.10
 CURLVERSION_FOURTH              7.16.1
+CURLVERSION_NINETH              7.75.0
```
Comment (jay) [inline docs/libcurl/symbols-in-versions:870]:
```suggestion
CURLVERSION_NINTH              7.75.0
```

Comment (bagder) [reply docs/libcurl/symbols-in-versions:870]: ack!

**Comment by bagder** [comment]
I've cleaned up the commits now in preparation for the first merge of this work. My plan with this merge is to provide a first version that users can check out and try (opt-in of course), but to not affect the default build at all.

I will then continue fixing the Hyper use further in follow-up PRs, hopefully much smaller and less disruptive.

**Pull request closed without merging**